lib/marked.js

```diff
@@ -7,8 +7,223 @@
 ;(function(root) {
 'use strict';
 
+/**
+ * Begin Marked
+ *
+ * TODO: None of our tests actually test a user just passing src?
+ *       `console.log(marked.receivedOnlySource(src, opt, callback));`
+ */
+
+/**
+ * @param  src string
+ * @param  opt (optional) Object
+ * @param  callback (optional) Object
+ *
+ * @return string
+ */
+function marked(src, opt, callback) {
+  // ^^ Can't we do something like marked(src, opt = {}, callback = function())
+
+  try {
+    marked.receivedValidSource(src);
+  } catch (e) {
+    throw e;
+  }
+
+  console.log(marked.receivedOnlySource(src, opt, callback));
+
+  // We essentially have two `marked`...this one conditional, and the one at the end...!!
+  if (marked.receivedCallback(opt, callback)) {
+    // TODO: This seems like an odd check, think there's a more readable way to
+    //       deal with optional params.
+    //
+    // We have a callback but not as the last param passed in.
+    if (!callback) {
+      callback = opt;
+      opt = null;
+    }
+
+    // If we don't have a callback honestly passed in, opt becomes null,
+    // then we set it to a merged of the defalts + opt, or an empty object??
+    opt = merge({}, marked.defaults, opt || {});
+
+    // Not sure what all is happening here.
+    var highlight = opt.highlight,
+        tokens,
+        pending,
+        i = 0;
```

Comment:
[According to MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator), the comma operator evaluates each operand and returns the value of the last operand.  Here, `opt.highlight`, `tokens`, `pending`, and `i = 0` are all being evaluated and (in theory) `0` is bing returned.  Intuitively (without testing) it appears that `highlight` is being set to `0` in a very convoluted way.  Maybe there's more to it.

Reply:
Concur with the assessment of the commenter. Wonder if we can add that to the lint (@UziTech)? Don't do this. :)

Reply:
I agree, I hate this style, but that is how chjj did it. I prefer initializing the variables as close to when you use them as possible.

In this case since the `=` operator has a [higher presedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#Table) than the `,` operator the expression is essentially:

```js
var (highlight = opt.highlight),
    (tokens),
    (pending),
    (i = 0);
```

Reply:
To be fair, I don't know who necessarily did it that way, just that I think it should be changed. Totally agree that putting things as close to their contextual relevance is kind of a big deal.

Reply:
There actually seems to be a lot of "clever" decisions that were made, which may not actually bring the desired ROI compared to writing human readable code. We spend way more time reading than we do writing, you know?